File: javascript/packages/linter/src/rules/erb-no-unused-literals.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"
import { PrismVisitor, PrismNodes } from "@herb-tools/core"

import { isERBOutputNode } from "@herb-tools/core"
import { locationFromOffset } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult, ERBContentNode, ParserOptions, PrismNode } from "@herb-tools/core"

class LiteralCollector extends PrismVisitor {
  public readonly literals: PrismNode[] = []

  override visit(node: PrismNode): void {
    if (!node) return

    super.visit(node)
  }

  visitCallNode(node: PrismNode): void {
    this.visit(node.receiver) // Only visit the receiver of a call, not its arguments.
  }

  // Stop traversal into all assignment node types.
  visitLocalVariableWriteNode(): void {}
  visitLocalVariableAndWriteNode(): void {}
  visitLocalVariableOperatorWriteNode(): void {}
  visitLocalVariableOrWriteNode(): void {}
  visitInstanceVariableWriteNode(): void {}
  visitInstanceVariableAndWriteNode(): void {}
  visitInstanceVariableOperatorWriteNode(): void {}
  visitInstanceVariableOrWriteNode(): void {}
  visitClassVariableWriteNode(): void {}
  visitClassVariableAndWriteNode(): void {}
  visitClassVariableOperatorWriteNode(): void {}
  visitClassVariableOrWriteNode(): void {}
  visitGlobalVariableWriteNode(): void {}
  visitGlobalVariableAndWriteNode(): void {}
  visitGlobalVariableOperatorWriteNode(): void {}
  visitGlobalVariableOrWriteNode(): void {}
  visitConstantWriteNode(): void {}
  visitConstantAndWriteNode(): void {}
  visitConstantOperatorWriteNode(): void {}
  visitConstantOrWriteNode(): void {}
  visitConstantPathWriteNode(): void {}
  visitConstantPathAndWriteNode(): void {}
  visitConstantPathOperatorWriteNode(): void {}
  visitConstantPathOrWriteNode(): void {}
  visitIndexAndWriteNode(): void {}
  visitIndexOperatorWriteNode(): void {}
  visitIndexOrWriteNode(): void {}
  visitCallAndWriteNode(): void {}
  visitCallOperatorWriteNode(): void {}
  visitCallOrWriteNode(): void {}
  visitMultiWriteNode(): void {}
  visitMatchWriteNode(): void {}

  visitArrayNode(node: PrismNodes.ArrayNode): void {
    this.literals.push(node)
  }

  visitFalseNode(node: PrismNodes.FalseNode): void {
    this.literals.push(node)
  }

  visitFloatNode(node: PrismNodes.FloatNode): void {
    this.literals.push(node)
  }

  visitHashNode(node: PrismNodes.HashNode): void {
    this.literals.push(node)
  }

  visitImaginaryNode(node: PrismNodes.ImaginaryNode): void {
    this.literals.push(node)
  }

  visitIntegerNode(node: PrismNodes.IntegerNode): void {
    this.literals.push(node)
  }

  visitNilNode(node: PrismNodes.NilNode): void {
    this.literals.push(node)
  }

  visitRangeNode(node: PrismNodes.RangeNode): void {
    this.literals.push(node)
  }

  visitRationalNode(node: PrismNodes.RationalNode): void {
    this.literals.push(node)
  }

  visitRegularExpressionNode(node: PrismNodes.RegularExpressionNode): void {
    this.literals.push(node)
  }

  visitStringNode(node: PrismNodes.StringNode): void {
    this.literals.push(node)
  }

  visitInterpolatedStringNode(node: PrismNodes.InterpolatedStringNode): void {
    this.literals.push(node)
  }

  visitSymbolNode(node: PrismNodes.SymbolNode): void {
    this.literals.push(node)
  }

  visitTrueNode(node: PrismNodes.TrueNode): void {
    this.literals.push(node)
  }
}

class ERBNoUnusedLiteralsVisitor extends BaseRuleVisitor {
  visitERBContentNode(node: ERBContentNode): void {
    if (isERBOutputNode(node)) return

    const prismNode = node.prismNode
    if (!prismNode) return

    const source = node.source
    if (!source) return

    const collector = new LiteralCollector()
    collector.visit(prismNode)

    for (const literal of collector.literals) {
      const { startOffset, length } = literal.location
      const literalSource = source.substring(startOffset, startOffset + length)
      const location = locationFromOffset(source, startOffset, length)

      this.addOffense(
        `Avoid using silent ERB tags for literals. \`${literalSource}\` is evaluated but never used or output.`,
        location,
        undefined,
        undefined,
        ["unnecessary"],
      )
    }
  }
}

export class ERBNoUnusedLiteralsRule extends ParserRule {
  static ruleName = "erb-no-unused-literals"
  static introducedIn = this.version("0.9.3")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: {
        cli: "error",
        editor: "info",
      }
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ERBNoUnusedLiteralsVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/erb_no_unused_literals.rs
use crate::offense::UnboundOffense;
use crate::rule::{LintContext, ParserRule, Rule};
use crate::utils::erb_utils::is_output_tag_opening;
use crate::utils::source_slice::location_from_offset;

use herb::nodes::ERBContentNode;
use herb::prism::PrismNode;
use herb::{ParseResult, Visitor};
use herb_config::{Severity, SeverityConfig};

const LITERAL_NODES: &[&str] = &[
  "ArrayNode",
  "FalseNode",
  "FloatNode",
  "HashNode",
  "ImaginaryNode",
  "IntegerNode",
  "NilNode",
  "RangeNode",
  "RationalNode",
  "RegularExpressionNode",
  "StringNode",
  "InterpolatedStringNode",
  "SymbolNode",
  "TrueNode",
];

/// Traversal stops at assignments and at control flow where literals are used
/// as return or flow values.
const OPAQUE_NODES: &[&str] = &["ReturnNode", "BreakNode", "NextNode", "MultiWriteNode", "MatchWriteNode"];

pub struct ERBNoUnusedLiteralsRule;

struct LiteralCollector<'node> {
  literals: Vec<&'node PrismNode>,
}

impl<'node> LiteralCollector<'node> {
  fn visit(&mut self, node: &'node PrismNode) {
    if node.node_type.ends_with("WriteNode") || OPAQUE_NODES.contains(&node.node_type.as_str()) {
      return;
    }

    if LITERAL_NODES.contains(&node.node_type.as_str()) {
      self.literals.push(node);
      return;
    }

    // the case operand and the `when` values are matched against, not discarded
    if node.is("CaseNode") || node.is("CaseMatchNode") {
      for condition in node.field("conditions") {
        self.visit(condition);
      }

      for clause in node.field("else_clause") {
        self.visit(clause);
      }

      return;
    }

    if node.is("WhenNode") || node.is("InNode") {
      for statement in node.field("statements") {
        self.visit(statement);
      }

      return;
    }

    // only the receiver of a call is visited, not its arguments
    if node.is("CallNode") {
      if let Some(receiver) = node.receiver() {
        self.visit(receiver);
      }

      return;
    }

    for child in &node.children {
      self.visit(child);
    }
  }
}

struct ERBNoUnusedLiteralsVisitor<'rule> {
  rule_name: &'static str,
  offenses: Vec<UnboundOffense>,
  source: &'rule str,
}

impl<'rule> Visitor for ERBNoUnusedLiteralsVisitor<'rule> {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let tag_opening = node.tag_opening.as_ref().map(|token| token.value.as_str()).unwrap_or("");

    if is_output_tag_opening(tag_opening) {
      return;
    }

    let prism_node = match node.prism() {
      Some(prism_node) => prism_node,
      None => return,
    };

    if self.source.is_empty() {
      return;
    }

    let mut collector = LiteralCollector { literals: Vec::new() };

    collector.visit(prism_node);

    for literal in collector.literals {
      let literal_source = self.source.get(literal.start_offset..literal.end_offset).unwrap_or("");

      self.offenses.push(UnboundOffense::with_tags(
        self.rule_name,
        format!(
          "Avoid using silent ERB tags for literals. `{}` is evaluated but never used or output.",
          literal_source
        ),
        location_from_offset(self.source, literal.start_offset, literal.end_offset),
        vec!["unnecessary".to_string()],
      ));
    }
  }
}

impl Rule for ERBNoUnusedLiteralsRule {
  fn name(&self) -> &'static str {
    "erb-no-unused-literals"
  }

  fn introduced_in(&self) -> Option<&'static str> {
    Some("0.9.3")
  }

  fn default_severity(&self) -> SeverityConfig {
    SeverityConfig::PerMode {
      cli: Severity::Error,
      editor: Severity::Info,
    }
  }

  fn parser_options(&self) -> herb::ParserOptions {
    herb::ParserOptions {
      prism_nodes: true,
      ..crate::rule::default_linter_parser_options()
    }
  }
}

impl ParserRule for ERBNoUnusedLiteralsRule {
  fn check(&self, result: &ParseResult, context: &LintContext) -> Vec<UnboundOffense> {
    let source = if context.source.is_empty() { &result.source } else { &context.source };

    let mut visitor = ERBNoUnusedLiteralsVisitor {
      rule_name: self.name(),
      offenses: Vec::new(),
      source,
    };

    visitor.visit_document_node(&result.value);

    visitor.offenses
  }
}
